### PR TITLE
Mobile app changes

### PR DESCRIPTION
**Pull request by ParryLuke**

### CODE REVIEW

File: api.swagger.yaml [hunk 1]
```diff
@@ -359,6 +443,31 @@ paths:
             schema:
               $ref: '#/components/schemas/user-schema'
         required: true
+  '/users/{user_id}/activity':
+    get:
+      tags:
+        - users
+      description: Return the user's activity feed.
+      parameters:
```
Comment (jmorahan) [inline api.swagger.yaml]: I think this needs a pager, or if not we should at least define how many items will be returned

File: api.swagger.yaml [hunk 2]
```diff
@@ -467,96 +576,803 @@ paths:
...
+        '400':
+          description: One or both of the UUIDs was malformed.
+        '404':
+          description: The user or location does not exist.
+  '/users/{user_id}/teams':
+    get:
+      tags:
+        - users
+      description: List a user's teams
+      parameters:
```
Comment (jmorahan) [inline api.swagger.yaml]: I think this should probably have a pager too...

File: api.swagger.yaml [hunk 3]
```diff
@@ -467,96 +576,803 @@ paths:
...
+      responses:
+        '200':
+          description: >
+            List of teams of which the user is a member, indicating the
+            membership type (admin or member) for each team. If the user is not
+            a member of any teams, an empty array is returned.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/user-team-membership-list-schema'
```
Comment (jmorahan) [inline api.swagger.yaml]: Can we pull out the array to avoid the null bug?

File: api.swagger.yaml [hunk 4]
```diff
@@ -467,96 +576,803 @@ paths:
...
+        '400':
+          description: The supplied UUID was malformed.
+        '404':
+          description: The UUID does not correspond to an existing user.
+  '/users/{user_id}/following':
+    get:
+      tags:
+        - users
+      description: List of who a user is following.
+      parameters:
```
Comment (jmorahan) [inline api.swagger.yaml]: tbh we should probably add pagers to everything that returns an array

File: api.swagger.yaml [hunk 5]
```diff
@@ -467,96 +576,803 @@ paths:
...
+        '400':
+          description: The supplied UUID was malformed.
+        '404':
+          description: The UUID does not correspond to an existing user.
+  '/users/{user_id}/groups':
+    get:
+      tags:
+        - users
+      description: List a user's groups.
+      parameters:
```
Comment (jmorahan) [inline api.swagger.yaml]: another one that probably should be paged

File: api.swagger.yaml [hunk 6]
```diff
@@ -467,96 +576,803 @@ paths:
...
+        '400':
+          description: The UUID was malformed.
+        '404':
+          description: The user or follower does not exist.
+      description: Follow a user as another user.
+  /groups:
+    get:
+      tags:
+        - groups
+      parameters:
```
Comment (jmorahan) [inline api.swagger.yaml]: this either needs a pager, or external_id should be required

File: api.swagger.yaml [hunk 7]
```diff
@@ -467,96 +576,803 @@ paths:
...
+        '400':
+          description: The supplied UUID was malformed.
+        '404':
+          description: The UUID does not correspond to an existing group.
+      description: Get information about the group.
+  '/groups/{group_id}/members':
+    get:
+      tags:
+        - groups
+      parameters:
```
Comment (jmorahan) [inline api.swagger.yaml]: this one definitely needs a pager

File: api.swagger.yaml [hunk 8]
```diff
@@ -467,96 +576,803 @@ paths:
...
+            pattern: >-
+              ^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$
+      responses:
+        '200':
+          description: |
+            List of group members. If there are no members, an empty array.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/group-membership-list-schema'
```
Comment (jmorahan) [inline api.swagger.yaml]: can we pull out the array to avoid the null bug

File: api.swagger.yaml [hunk 9]
```diff
@@ -467,96 +576,803 @@ paths:
...
+              ^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$
+        - description: The Invotra UUID of the user.
+          in: path
+          name: user_id
+          required: true
+          schema:
+            type: string
+            pattern: >-
+              ^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$
+      responses:
```
Comment (jmorahan) [inline api.swagger.yaml]: Looking at this one it occurred to me that it may also need a 403 response for when you're neither the user nor a group admin. Then I thought maybe we need 403s for a few others, eg private/secret groups that you're not a member of. But maybe that should be a separate PR to add all the 403s

File: api.swagger.yaml [hunk 10]
```diff
@@ -467,96 +576,803 @@ paths:
...
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: '#/components/schemas/group-membership-common-schema'
+  /posts:
+    get:
+      tags:
+        - posts
+      parameters:
```
Comment (jmorahan) [inline api.swagger.yaml]: again the external_id should either be required or we need a pager

File: api.swagger.yaml [hunk 11]
```diff
@@ -467,96 +576,803 @@ paths:
...
+                The URL at which the newly provisioned post may be retrieved.
+              schema:
+                type: string
+        '400':
+          description: The post was not saved due to malformed JSON.
+        '422':
+          description: |
+            The post was not saved due to Drupal validation failure or JSON
+            schema validation failure.
+      description: Provision a new post.
```
Comment (jmorahan) [inline api.swagger.yaml]: I think "provision" is a strange word to use in this context :)

File: api.swagger.yaml [hunk 12]
```diff
@@ -467,96 +576,803 @@ paths:
...
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/user-schema'
+        '400':
+          description: The supplied UUID was malformed.
+        '404':
+          description: The UUID does not correspond to an existing post.
+      description: Get a list of likes for a post.
```
Comment (jmorahan) [inline api.swagger.yaml]: I think we should clarify what a "like" actually is - in this case, a user (who liked the post)

I know this can be inferred from the schema but no harm to describe it too :)

File: api.swagger.yaml [hunk 13]
```diff
@@ -1151,10 +2044,10 @@ components:
             https://demo.invotra.com/uuid/node/1f0c6cb5-a3c3-4197-bebc-b3e44462dc26
         short_text:
           description: >-
-            Contextual text for the feed message, with placeholders @user for the user's
-            name and optionally @group for the group name.
+            Contextual text for the feed message, which displays in between the
+            user's name and the entity title
           type: string
-          example: "@user created the following within @group"
+          example: created the following within
```
Comment (jmorahan) [inline api.swagger.yaml]: Did you mean to change these or was it a copy/paste error?

File: api.swagger.yaml [hunk 14]
```diff
@@ -1141,7 +2034,7 @@ components:
           type: string
           example: I'm taking annual leave next Thursday. Cover to be announced shortly
         timestamp:
-          description: This is timestamp (in seconds) of when the feed message was created
+          description: This is timestamp of when the feed message was created
```
Comment (jmorahan) [inline api.swagger.yaml]: Did you mean to change this or was it a copy/paste error?

File: api.swagger.yaml [hunk 15]
```diff
@@ -1201,7 +2106,7 @@ components:
           type: string
           example: 1f0c6cb5-a3c3-4197-bebc-b3e44462dc23
         timestamp:
-          description: This is timestamp (in seconds) of when the notification was created
+          description: This is timestamp of when the notificaiton was created
```
Comment (jmorahan) [inline api.swagger.yaml]: I assume this one was a copy/paste error as it also reintroduces a spelling error ;)

File: api.swagger.yaml [hunk 16]
```diff
@@ -1327,6 +2254,233 @@ components:
...
+          pattern: >-
+            ^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$
+          example: 01234567-89ab-cdef-1234-56789abcdef0
+        privacy_setting:
+          description: This is the privacy setting for the group.
+          type: string
+          enum:
+            - Public
+            - Private
+            - Secret
```
Comment (jmorahan) [inline api.swagger.yaml]: I wonder if these should just be lowercase? As in this context they're effectively machine names

File: api.swagger.yaml [hunk 17]
```diff
@@ -1327,6 +2254,233 @@ components:
...
+      type: object
+      properties:
+        filename:
+          description: This is the filename
+          type: string
+          example: wallpaper.jpeg
+        file_download:
+          description: The location of the file
+          type: string
+          example: 'file/1234/download'
```
Comment (jmorahan) [inline api.swagger.yaml]: Wouldn't this be a full URL with a protocol and host etc?

File: api.swagger.yaml [hunk 18]
```diff
@@ -242,7 +276,21 @@ paths:
         - description: Filter by external ID.
           in: query
           name: external_id
-          required: true
+          schema:
+            type: string
+        - description: Search term.
+          in: query
+          name: fulltext_search
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: Removed. Also reverted /users back to external_id required so that it doesn't change for Sierra.

File: api.swagger.yaml [hunk 19]
```diff
@@ -285,6 +333,42 @@ paths:
             schema:
               $ref: '#/components/schemas/user-schema-create'
         required: true
+  /users/search:
+    get:
+      tags:
+        - users
+      parameters:
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: added

File: api.swagger.yaml [hunk 20]
```diff
@@ -285,6 +333,42 @@ paths:
...
+        '200':
+          description: List of users matching the filter criteria.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/user-search-schema'
+        '404':
+          description: No matching users were found.
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml [hunk 21]
```diff
@@ -359,6 +443,31 @@ paths:
...
+            type: string
+      responses:
+        '200':
+          description: >
+            Activity feed returned successfully. An empty array is returned if
+            there are no results.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/feed-list-schema'
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml [hunk 22]
```diff
@@ -467,96 +576,803 @@ paths:
...
+      parameters:
+        - description: The Invotra UUID of the user.
+          in: path
+          name: user_id
+          required: true
+          schema:
+            type: string
+            pattern: >-
+              ^[a-fA-F0-9]{8}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{4}-[a-fA-F0-9]{12}$
+        - description: The Invotra UUID of the job role.
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 1 above)
Comment (ParryLuke) [reply api.swagger.yaml]: added

File: api.swagger.yaml (hunk 2 above)
Comment (ParryLuke) [reply api.swagger.yaml]: added

File: api.swagger.yaml (hunk 3 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 4 above)
Comment (ParryLuke) [reply api.swagger.yaml]: added - I think I got all of them

File: api.swagger.yaml (hunk 5 above)
Comment (ParryLuke) [reply api.swagger.yaml]: added

File: api.swagger.yaml (hunk 6 above)
Comment (ParryLuke) [reply api.swagger.yaml]: added

File: api.swagger.yaml (hunk 7 above)
Comment (ParryLuke) [reply api.swagger.yaml]: added

File: api.swagger.yaml (hunk 8 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 9 above)
Comment (ParryLuke) [reply api.swagger.yaml]: Will do in a separate PR

File: api.swagger.yaml (hunk 10 above)
Comment (ParryLuke) [reply api.swagger.yaml]: added

File: api.swagger.yaml (hunk 11 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 12 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 13 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 14 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 15 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 16 above)
Comment (ParryLuke) [reply api.swagger.yaml]: agreed/updated

File: api.swagger.yaml (hunk 17 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml [hunk 23]
```diff
@@ -1327,6 +2242,229 @@ components:
...
+            example: 5
+          attachments_count:
+            description: Count of attachments on the post
+            type: integer
+            example: 5
+    post-schema-create:
+      allOf:
+      - $ref: '#/components/schemas/post-schema-base'
+      - type: object
+        properties:
```
Comment (jmorahan) [inline api.swagger.yaml]: I wonder should we also be using a uuid for the category here, rather than the name? for create and update

File: api.swagger.yaml [hunk 24]
```diff
@@ -473,90 +580,786 @@ paths:
...
+            Like status returned
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/common-status-schema'
+        '400':
+          description: The UUID was malformed.
+        '404':
+          description: The post does not exist. 
+    description: Check if the current user has liked a post.
```
Comment (jmorahan) [inline api.swagger.yaml]: this seems to be indented wrong

File: api.swagger.yaml [hunk 25]
```diff
@@ -680,6 +1483,28 @@ paths:
...
+        - $ref: '#/components/parameters/limitParam'
+      responses:
+        '200':
+          description: List of locations matching the filter criteria.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/location-uuid-label-list-schema'
+        '404':
+          description: No matching locations were found.
```
Comment (jmorahan) [inline api.swagger.yaml]: Another 404 that should be an empty array! No idea how I missed all of these on the previous review :(

File: api.swagger.yaml (hunk 23 above)
Comment (ParryLuke) [reply api.swagger.yaml]: Updated - for now the GET returns a string and the UUID because the endpoints are sharing the base schema.

I could change it so that get, post and put all use 3 different schemas?

File: api.swagger.yaml [hunk 26]
```diff
@@ -473,90 +580,786 @@ paths:
...
+        - $ref: '#/components/parameters/limitParam'
+      responses:
+        '200':
+          description: List of group categories.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/title-uuid-list-common-schema'
+        '404':
+          description: No group categories were found.
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml [hunk 27]
```diff
@@ -473,90 +580,786 @@ paths:
...
+        '200':
+          description: List of groups matching the filter criteria.
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: '#/components/schemas/group-search-schema'
+        '404':
+          description: No matching groups were found.
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml [hunk 28]
```diff
@@ -1079,7 +1920,50 @@ components:
...
+      description: The numbers of items to return.
   schemas:
+    branding-schema:
+      properties:
+        toolbar_colour:
+          description: The colour for the toolbar.
+          type: string
+          example: '#FFFFFF'
+        toolbar_icons_colour:
+          description: This is colour for hyperlinks.
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: yep :)

File: api.swagger.yaml [hunk 29]
```diff
@@ -86,6 +98,28 @@ paths:
...
+        - $ref: '#/components/parameters/limitParam'
+      responses:
+        '200':
+          description: List of teams matching the filter criteria.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/uuid-label-list-schema'
+        '404':
+          description: No matching teams were found.
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: updated. Also updated /teams/list and /locations/list

File: api.swagger.yaml [hunk 30]
```diff
@@ -1151,8 +2032,8 @@ components:
             https://demo.invotra.com/uuid/node/1f0c6cb5-a3c3-4197-bebc-b3e44462dc26
         short_text:
           description: >-
-            Contextual text for the feed message, with placeholders @user for the user's
-            name and optionally @group for the group name.
+            Contextual text for the feed message, which displays in between the
+            user's name and the entity title
```
[the remark this answers is not in the document]
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 24 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated

File: api.swagger.yaml (hunk 25 above)
Comment (ParryLuke) [reply api.swagger.yaml]: updated
